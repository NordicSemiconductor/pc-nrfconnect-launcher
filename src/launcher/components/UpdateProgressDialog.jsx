/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import React from 'react';
import Button from 'react-bootstrap/Button';
import Modal from 'react-bootstrap/Modal';
import ProgressBar from 'react-bootstrap/ProgressBar';
import { Spinner } from 'pc-nrfconnect-shared';
import { bool, func, number, string } from 'prop-types';

const UpdateProgressDialog = ({
    isVisible,
    isProgressSupported,
    isCancelSupported,
    version,
    percentDownloaded,
    onCancel,
    isCancelling,
}) => (
    <Modal show={isVisible} backdrop>
        <Modal.Header closeButton={false}>
            <Modal.Title>Downloading update</Modal.Title>
        </Modal.Header>
        <Modal.Body>
            <p>Downloading nRF Connect {version}...</p>
            {isProgressSupported && (
                <ProgressBar
                    label={`${percentDownloaded}%`}
                    now={percentDownloaded}
                />
            )}
            <p>
                This might take a few minutes. The application will restart and
                update once the download is complete.
            </p>
        </Modal.Body>
        <Modal.Footer>
            {!isProgressSupported && <Spinner />}
            {isCancelSupported && (
                <Button
                    onClick={onCancel}
                    disabled={isCancelling || percentDownloaded === 100}
                >
                    Cancel
                </Button>
            )}
        </Modal.Footer>
    </Modal>
);

UpdateProgressDialog.propTypes = {
    isVisible: bool.isRequired,
    isProgressSupported: bool.isRequired,
    isCancelSupported: bool.isRequired,
    version: string.isRequired,
    percentDownloaded: number.isRequired,
    onCancel: func.isRequired,
    isCancelling: bool.isRequired,
};

export default UpdateProgressDialog;
