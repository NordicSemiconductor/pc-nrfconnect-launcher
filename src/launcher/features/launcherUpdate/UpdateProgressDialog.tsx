/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import ProgressBar from 'react-bootstrap/ProgressBar';
import { Dialog, DialogButton, Spinner } from 'pc-nrfconnect-shared';

import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { cancelDownload } from './launcherUpdateEffects';
import { getLauncherUpdate } from './launcherUpdateSlice';

export default () => {
    const dispatch = useLauncherDispatch();
    const {
        isUpdateProgressDialogVisible: isVisible,
        isProgressSupported,
        isCancelSupported,
        latestVersion: version,
        percentDownloaded,
        isCancelling,
    } = useLauncherSelector(getLauncherUpdate);

    return (
        <Dialog isVisible={isVisible} closeOnUnfocus={false}>
            <Dialog.Header title="Downloading update" />
            <Dialog.Body>
                <p>Downloading nRF Connect for Desktop {version}...</p>
                {isProgressSupported && (
                    <ProgressBar
                        label={`${percentDownloaded}%`}
                        now={percentDownloaded}
                    />
                )}
                <p>
                    This might take a few minutes. The application will restart
                    and update once the download is complete.
                </p>
            </Dialog.Body>
            <Dialog.Footer>
                {!isProgressSupported && <Spinner />}
                {isCancelSupported && (
                    <DialogButton
                        onClick={() => dispatch(cancelDownload())}
                        disabled={isCancelling || percentDownloaded === 100}
                        variant="primary"
                    >
                        Cancel
                    </DialogButton>
                )}
            </Dialog.Footer>
        </Dialog>
    );
};
