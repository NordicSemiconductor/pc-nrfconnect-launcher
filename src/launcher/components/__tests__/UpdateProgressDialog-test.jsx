/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/* eslint-disable import/first */

// Do not render react-bootstrap components in tests
jest.mock('react-bootstrap', () => ({
    Modal: 'Modal',
    Button: 'Button',
    ModalHeader: 'ModalHeader',
    ModalFooter: 'ModalFooter',
    ModalBody: 'ModalBody',
    ModalTitle: 'ModalTitle',
    ProgressBar: 'ProgressBar',
}));

import React from 'react';
import { shallow } from 'enzyme';

import UpdateProgressDialog from '../UpdateProgressDialog';

describe('UpdateProgressDialog', () => {
    it('should render invisible', () => {
        expect(
            shallow(
                <UpdateProgressDialog
                    isVisible={false}
                    isProgressSupported={false}
                    isCancelSupported={false}
                    version=""
                    percentDownloaded={0}
                    onCancel={() => {}}
                    isCancelling={false}
                />
            )
        ).toMatchSnapshot();
    });

    it('should render with version, percent downloaded, and cancellable', () => {
        expect(
            shallow(
                <UpdateProgressDialog
                    isVisible
                    isProgressSupported
                    isCancelSupported
                    version="1.2.3"
                    percentDownloaded={42}
                    onCancel={() => {}}
                    isCancelling={false}
                />
            )
        ).toMatchSnapshot();
    });

    it('should render with version, without progress, and not cancellable', () => {
        expect(
            shallow(
                <UpdateProgressDialog
                    isVisible
                    isProgressSupported={false}
                    isCancelSupported={false}
                    version="1.2.3"
                    percentDownloaded={0}
                    onCancel={() => {}}
                    isCancelling={false}
                />
            )
        ).toMatchSnapshot();
    });

    it('should render cancelling', () => {
        expect(
            shallow(
                <UpdateProgressDialog
                    isVisible
                    isProgressSupported
                    isCancelSupported
                    version="1.2.3"
                    percentDownloaded={42}
                    onCancel={() => {}}
                    isCancelling
                />
            )
        ).toMatchSnapshot();
    });

    it('should render with cancel disabled when 100 percent complete', () => {
        expect(
            shallow(
                <UpdateProgressDialog
                    isVisible
                    isProgressSupported
                    isCancelSupported
                    version="1.2.3"
                    percentDownloaded={100}
                    onCancel={() => {}}
                    isCancelling={false}
                />
            )
        ).toMatchSnapshot();
    });
});
