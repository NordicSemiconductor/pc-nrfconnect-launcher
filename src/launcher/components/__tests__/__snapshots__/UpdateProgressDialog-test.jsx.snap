// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateProgressDialog should render cancelling 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "ModalDialog",
      "render": [Function],
    }
  }
  enforceFocus={true}
  keyboard={true}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    closeButton={false}
    closeLabel="Close"
  >
    <ModalTitle>
      Downloading update
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <p>
      Downloading nRF Connect for Desktop 
      1.2.3
      ...
    </p>
    <ProgressBar
      animated={false}
      isChild={false}
      label="42%"
      max={100}
      min={0}
      now={42}
      srOnly={false}
      striped={false}
    />
    <p>
      This might take a few minutes. The application will restart and update once the download is complete.
    </p>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      disabled={true}
      onClick={[Function]}
      variant="primary"
    >
      Cancel
    </Button>
  </ModalFooter>
</Modal>
`;

exports[`UpdateProgressDialog should render invisible 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "ModalDialog",
      "render": [Function],
    }
  }
  enforceFocus={true}
  keyboard={true}
  restoreFocus={true}
  show={false}
>
  <ModalHeader
    closeButton={false}
    closeLabel="Close"
  >
    <ModalTitle>
      Downloading update
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <p>
      Downloading nRF Connect for Desktop 
      ...
    </p>
    <p>
      This might take a few minutes. The application will restart and update once the download is complete.
    </p>
  </ModalBody>
  <ModalFooter>
    <_default />
  </ModalFooter>
</Modal>
`;

exports[`UpdateProgressDialog should render with cancel disabled when 100 percent complete 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "ModalDialog",
      "render": [Function],
    }
  }
  enforceFocus={true}
  keyboard={true}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    closeButton={false}
    closeLabel="Close"
  >
    <ModalTitle>
      Downloading update
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <p>
      Downloading nRF Connect for Desktop 
      1.2.3
      ...
    </p>
    <ProgressBar
      animated={false}
      isChild={false}
      label="100%"
      max={100}
      min={0}
      now={100}
      srOnly={false}
      striped={false}
    />
    <p>
      This might take a few minutes. The application will restart and update once the download is complete.
    </p>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      disabled={true}
      onClick={[Function]}
      variant="primary"
    >
      Cancel
    </Button>
  </ModalFooter>
</Modal>
`;

exports[`UpdateProgressDialog should render with version, percent downloaded, and cancellable 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "ModalDialog",
      "render": [Function],
    }
  }
  enforceFocus={true}
  keyboard={true}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    closeButton={false}
    closeLabel="Close"
  >
    <ModalTitle>
      Downloading update
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <p>
      Downloading nRF Connect for Desktop 
      1.2.3
      ...
    </p>
    <ProgressBar
      animated={false}
      isChild={false}
      label="42%"
      max={100}
      min={0}
      now={42}
      srOnly={false}
      striped={false}
    />
    <p>
      This might take a few minutes. The application will restart and update once the download is complete.
    </p>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      variant="primary"
    >
      Cancel
    </Button>
  </ModalFooter>
</Modal>
`;

exports[`UpdateProgressDialog should render with version, without progress, and not cancellable 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "ModalDialog",
      "render": [Function],
    }
  }
  enforceFocus={true}
  keyboard={true}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    closeButton={false}
    closeLabel="Close"
  >
    <ModalTitle>
      Downloading update
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <p>
      Downloading nRF Connect for Desktop 
      1.2.3
      ...
    </p>
    <p>
      This might take a few minutes. The application will restart and update once the download is complete.
    </p>
  </ModalBody>
  <ModalFooter>
    <_default />
  </ModalFooter>
</Modal>
`;
